pkgs/by-name/co/cosmic-term/package.nix

```
@@ -0,0 +1,97 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  rust,
+  rustPlatform,
+  cmake,
+  makeBinaryWrapper,
+  cosmic-icons,
+  just,
+  pkg-config,
+  libxkbcommon,
+  glib,
+  gtk3,
+  libinput,
+  fontconfig,
+  freetype,
+  wayland,
+  xorg,
+}:
+
+rustPlatform.buildRustPackage rec {
+  pname = "cosmic-term";
+  version = "unstable-2023-12-26";
+
+  src = fetchFromGitHub {
+    owner = "pop-os";
+    repo = pname;
+    rev = "bf3f507fdd73a06ab1f9b199a98dca6988aafec2";
+    hash = "sha256-c5RNrC0AZvz+O3nj7VvMQuA/U0sgxZCVHn+cc+4pIN8=";
+  };
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHashes = {
+      "accesskit-0.11.0" = "sha256-xVhe6adUb8VmwIKKjHxwCwOo5Y1p3Or3ylcJJdLDrrE=";
+      "atomicwrites-0.4.2" = "sha256-QZSuGPrJXh+svMeFWqAXoqZQxLq/WfIiamqvjJNVhxA=";
+      "cosmic-config-0.1.0" = "sha256-V371fmSmLIwUxtx6w+C55cBJ8oyYgN86r3FZ8rGBLEs=";
+      "cosmic-text-0.10.0" = "sha256-/4Hg+7R0LRF4paXIREkMOTtbQ1xgONym5nKb/TuyeD4=";
+      "glyphon-0.3.0" = "sha256-T7hvqtR3zi9wNemFrPPGakq2vEraLpnPkN7umtumwVg=";
+      "sctk-adwaita-0.5.4" = "sha256-yK0F2w/0nxyKrSiHZbx7+aPNY2vlFs7s8nu/COp2KqQ=";
+      "softbuffer-0.3.3" = "sha256-eKYFVr6C1+X6ulidHIu9SP591rJxStxwL9uMiqnXx4k=";
+      "smithay-client-toolkit-0.16.1" = "sha256-z7EZThbh7YmKzAACv181zaEZmWxTrMkFRzP0nfsHK6c=";
+      "taffy-0.3.11" = "sha256-SCx9GEIJjWdoNVyq+RZAGn0N71qraKZxf9ZWhvyzLaI=";
+      "winit-0.28.6" = "sha256-FhW6d2XnXCGJUMoT9EMQew9/OPXiehy/JraeCiVd76M=";
+    };
+  };
+
+  postPatch = ''
+    substituteInPlace justfile --replace '#!/usr/bin/env' "#!$(command -v env)"
+  '';
+
+  nativeBuildInputs = [
+    cmake
+    just
+    pkg-config
+    makeBinaryWrapper
+  ];
+  buildInputs = [
+    libxkbcommon
+    xorg.libX11
+    libinput
+    fontconfig
+    freetype
+    wayland
+    glib
+    gtk3
+  ];
+
+  dontUseJustBuild = true;
+
+  justFlags = [
+    "--set"
+    "prefix"
+    (placeholder "out")
+    "--set"
+    "bin-src"
+    "target/${
+      rust.lib.toRustTargetSpecShort stdenv.hostPlatform
+    }/release/cosmic-term"
+  ];
+
+  # LD_LIBRARY_PATH can be removed once tiny-xlib is bumped above 0.2.2
+  postInstall = ''
+    wrapProgram "$out/bin/${pname}" \
+      --suffix XDG_DATA_DIRS : "${cosmic-icons}/share" \
+      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ xorg.libX11 ]}
```

You'll need to add two extra dependencies to LD_LIBRARY_PATH because winit dlopens them on wayland:
```suggestion
      --prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath [ xorg.libX11 wayland libxkbcommon ]}
```

BTW, since cosmic-epoch is wayland-only, do we still need to keep X11?

---

I agree, X11 should be dropped where possible or XWayland should be used if X11 compatibility is needed 🤔

---

It builds without them is there a reason to add them?

The X11 packages will need to stay there to keep it working on Xorg.

---

This is the same as cosmic-edit:

https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/co/cosmic-edit/package.nix#L87

---

> I agree, X11 should be dropped where possible or XWayland should be used if X11 compatibility is needed 🤔 (I wasn't aware that it was Wayland only)

By wayland only, I mean that there's no X11 session anymore -- the session is wayland only. Xwayland still exists to keep backward compatibility.

> It builds without them is there a reason to add them?

It will give a `NoWaylandLib` error if you unset `DISPLAY`, since it tries to use `dlopen` to open the wayland library, which is not existent on NixOS.

To be able to find the library, you'll either to force linking like this:

https://github.com/NixOS/nixpkgs/blob/d1fcabefe1617c4dd295774692140b2018b9f9fc/pkgs/by-name/co/cosmic-panel/package.nix#L47C2-L53C7

or wrap with LD_LIBRARY_PATH including wayland.

> The X11 packages will need to stay there to keep it working on Xorg.

If you're using cosmic-term with cosmic-session, then there's no reason to keep X11. Although one could argue that they want to use cosmic tools on some other X11 DE, in which case yes the package needs to be kept.

---

This is meant to go with cosmic-session so you would have `DISPLAY` set so I believe it is best to have it set. Is it not up to the package/software developer to pick the deps for a package?

---

I unset `DISPLAY` simply to test if it works under wayland. Currently without adding `wayland` and `libxkbcommon` to the `LD_LIBRARY_PATH` it will refuse to run under wayland at all. It's not ideal to have to run through Xwayland for an application that is designed to work in a Wayland-only DE.

---

Wayland of `cosmic-edit` works because it links directly with libwayland-egl instead of opening it via dlopen. That's not true for `cosmic-term`.

---

Updated it and it seems to work on X11/Xorg for me still. Does it have any issues on Wayland?